xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -0,0 +1,652 @@
+
+// Reciprocal of the square root of variance + epsilon with optional broadcast.
+auto NormFactor(HloInstruction** norm_factor, HloInstruction** input0,
+                HloInstruction** input1, HloInstruction** variance,
+                HloInstruction** expectation, HloInstruction** epsilon) {
+  auto shared_subpattern = m::SharedSubpattern(Rsqrt(
+      norm_factor, AddAnyOrder(Variance(variance, expectation, input0, input1),
+                               m::Broadcast(m::ConstantScalar(epsilon)))));
+  return m::AnyOf<HloInstruction>(m::Broadcast(shared_subpattern),
+                                  shared_subpattern);
+}
+
+// Any order of p0 * p1 * p2.
+template <typename P0, typename P1, typename P2>
+auto MultiplyMultiplyAnyOrder(P0 p0, P1 p1, P2 p2) {
+  return m::AnyOf<HloInstruction>(
+      MultiplyAnyOrder(p0, MultiplyAnyOrder(p1, p2)),
+      MultiplyAnyOrder(p1, MultiplyAnyOrder(p0, p2)),
+      MultiplyAnyOrder(p2, MultiplyAnyOrder(p0, p1)));
+}
+
+// Any order of p0 - p1 + p2.
+template <typename P0, typename P1, typename P2>
+auto SubtractAddAnyOrder(P0 p0, P1 p1, P2 p2) {
+  return m::AnyOf<HloInstruction>(m::AddAnyOrder(m::Subtract(p0, p1), p2),
+                                  m::AddAnyOrder(m::Subtract(p2, p1), p0),
+                                  m::Subtract(m::AddAnyOrder(p0, p2), p1));
+}
+
+// Any order of (p0 - p1) * p2 * p3 + p4.
+template <typename P0, typename P1, typename P2, typename P3, typename P4>
+auto SubtractMultiplyAddAnyOrder(P0 p0, P1 p1, P2 p2, P3 p3, P4 p4) {
+  return m::AnyOf<HloInstruction>(
+      SubtractAddAnyOrder(MultiplyMultiplyAnyOrder(p0, p2, p3),
+                          MultiplyMultiplyAnyOrder(p1, p2, p3), p4),
+      m::AddAnyOrder(MultiplyMultiplyAnyOrder(Subtract(p0, p1), p2, p3), p4));
+}
+
+class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
+ public:
+  explicit CudnnNormRewriterVisitor(
+      const se::CudaComputeCapability cuda_compute_capability)
+      : cuda_compute_capability_(cuda_compute_capability) {}
+
+  Status HandleAdd(HloInstruction* instr) override {
+    return MatchLayerNorm(instr);
+  }
+
+  Status HandleSubtract(HloInstruction* instr) override {
+    return MatchLayerNorm(instr);
+  }
+
+  // Matches and rewrites a layer norm pattern with scale multiplication and
+  // bias addition into a Custom Call to cuDNN.
+  Status MatchLayerNorm(HloInstruction* instr) {
+    HloInstruction *input, *input0, *input1, *input2, *scale, *bias, *epsilon,
+        *expectation, *expectation0, *reduce, *norm_factor, *variance;
+    if (Match(instr,
+              SubtractMultiplyAddAnyOrder(
+                  m::Op(&input),
+                  Expectation(&expectation, &reduce, m::Op(&input0)),
+                  NormFactor(&norm_factor, &input1, &input2, &variance,
+                             &expectation0, &epsilon),
+                  // NormFactor(&input1, &input2, &epsilon),
+                  m::Broadcast(m::Op(&scale)), m::Broadcast(m::Op(&bias))))) {
```

Comment:
Check that scale is broadcasted in the right dimension. This might be slightly tricky as the reshapes can change the effective normalized dimensions. Maybe add a test for the case where scale is broadcast in the wrong dimension.